### PR TITLE
Refactor Drawer

### PR DESCRIPTION
**Pull request by elParaguayo**
This is based on the wlroots0.16 branch (#3985) and makes some changes to tidy up the codebase,

### CODE REVIEW

File: libqtile/backend/drawer.py [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+    def clear(self, colour):
+        self.ctx.save()
+        self.ctx.set_operator(cairocffi.OPERATOR_CLEAR)
+        self.ctx.rectangle(0, 0, self.width, self.height)
+        self.ctx.fill()
+        self.ctx.set_operator(cairocffi.OPERATOR_SOURCE)
+        self.set_source_rgb(colour)
+        self.ctx.rectangle(0, 0, self.width, self.height)
+        self.ctx.fill()
+        self.ctx.restore()
```
Comment (m-col) [inline libqtile/backend/drawer.py:276]: How come these lines were added? Was the CLEAR first not enough?

**Review by m-col — COMMENTED** [review]
From what I could tell reviewing the PR, beside the `Bar` changes, the changes are mainly dropping the `clear` methods of the x11 and wayland `Drawer` classes, and adding a few lines to the `clear` method of the base `Drawer` method. The backend-specific implementations referenced backend specific surfaces (`self_xcb_surface` and `self.surface`). Are those actions replaced by the new lines in the base class `clear` implementation?

If things work then LGTM, I'm happy to append these changes to the wlroots PR.

**Comment by elParaguayo** [comment]
That's correct. As everything is now done on the recording surface, there's no need for a backend-specific implementation of `clear`.

**Comment by elParaguayo** [comment]
I think there may be a couple of other bits to tidy up (if I recall correctly, this might impact a couple of widgets too) and I should do a last check on widget decorations.

**Comment by elParaguayo** [comment]
@m-col I think this is ok for rebasing onto your branch now.

**Comment by elParaguayo** [comment]
Closing as changes have been added to #3985

**Pull request closed without merging**